### ISSUE TITLE
HTTP PUT with json containing percent fails

### ISSUE THREAD

**Issue by voelzmo** [sinatra/sinatra#1286]
## Steps to reproduce
* Create a sinatra app which accepts HTTP PUT
```
require 'sinatra'

put '/' do
  'you put some data!'
end
```

* run with `ruby myapp.rb`
* put some json data which contains a percent `%`
```
$ curl -XPUT -d '{"password": "%password"}' http://localhost:4567
```

## Expected
This just works

## Actual
Stacktrace
```
Rack::Utils::InvalidParameterError: invalid %-encoding ({"password": "%password"})
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-1.6.5/lib/rack/utils.rb:127:in `rescue in parse_nested_query'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-1.6.5/lib/rack/utils.rb:117:in `parse_nested_query'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-1.6.5/lib/rack/request.rb:371:in `parse_query'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-1.6.5/lib/rack/request.rb:215:in `POST'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-1.6.5/lib/rack/request.rb:230:in `params'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/sinatra-1.4.8/lib/sinatra/base.rb:902:in `call!'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/sinatra-1.4.8/lib/sinatra/base.rb:895:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-protection-1.5.3/lib/rack/protection/xss_header.rb:18:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-protection-1.5.3/lib/rack/protection/path_traversal.rb:16:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-protection-1.5.3/lib/rack/protection/json_csrf.rb:18:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-protection-1.5.3/lib/rack/protection/base.rb:49:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-protection-1.5.3/lib/rack/protection/base.rb:49:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-protection-1.5.3/lib/rack/protection/frame_options.rb:31:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-1.6.5/lib/rack/logger.rb:15:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-1.6.5/lib/rack/commonlogger.rb:33:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/sinatra-1.4.8/lib/sinatra/base.rb:219:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/sinatra-1.4.8/lib/sinatra/base.rb:212:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-1.6.5/lib/rack/head.rb:13:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-1.6.5/lib/rack/methodoverride.rb:22:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/sinatra-1.4.8/lib/sinatra/show_exceptions.rb:25:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/sinatra-1.4.8/lib/sinatra/base.rb:182:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/sinatra-1.4.8/lib/sinatra/base.rb:2013:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/sinatra-1.4.8/lib/sinatra/base.rb:1487:in `block in call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/sinatra-1.4.8/lib/sinatra/base.rb:1787:in `synchronize'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/sinatra-1.4.8/lib/sinatra/base.rb:1487:in `call'
	/Users/me/.rvm/gems/ruby-2.3.1@sinatra/gems/rack-1.6.5/lib/rack/handler/webrick.rb:88:in `service'
	/Users/me/.rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/webrick/httpserver.rb:140:in `service'
	/Users/me/.rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/webrick/httpserver.rb:96:in `run'
	/Users/me/.rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/webrick/server.rb:296:in `block in start_thread'
```

## Note
This works when using plain rack, therefore not sure how much it is related to https://github.com/rack/rack/issues/520 and https://github.com/rack/rack/issues/337 respectively.

Sample Rack app:
```
require 'rack'
require 'thin'
require 'json'

class RackApp
  def self.call(env)
    req = Rack::Request.new(env)
    p JSON.parse( req.body.read )
    Rack::Response.new("Hello, World!")
  end
end

Rack::Handler::Thin.run RackApp
```

Run it with `ruby rackapp.rb`

And the HTTP PUT:
```
$ curl -XPUT -d '{"password": "%password"}' http://localhost:8080
```
just works.

Any ideas?

**Comment by mwpastore** [comment]
This is more or less by design. Since you are sending a JSON-encoded payload, you should set the content-type of the request to application/json:

```console
$ curl -XPUT -d '{"password": "%password"}' -H 'Content-Type: application/json' http://localhost:4567
```

**Comment by mwpastore** [comment]
I will concede that there are some puzzling inconsistencies in how this is handled in Rack 1.6 and Sinatra 1.4 (e.g. PUT vs. POST, Rack vs. Sinatra), but they appear to have been addressed in Rack 2.0 and Sinatra 2.0 (beta/RC).

**Comment by voelzmo** [comment]
Ah, great, thanks!

I've verified it works with the header and sinatra `1.4.7`. Will update my client implementation.
You're right, the inconsistencies between PUT and POST was also very weird to me.
